library/Test/Hspec/JUnitFormatter.hs

```diff
@@ -0,0 +1,159 @@
+module Test.Hspec.JUnitFormatter
+  ( use
+  , register
+  , formatter
+  , formatterWith
+  , module Api
+  ) where
```

Comment:
So is the idea that formatter authors would provide a module with exactly this interface (specifically `use` and `register`) and not our `Hspec.JUnit.hs` module?

I ask because I think we can simplify `hspecJUnitWith` to something like,

```hs
hspecWith base
  (if junitEnabled then JUnitFormatter.use else JUnitFormatter.register)
  spec
```

And, at that point, the module's not particularly valuable.

Reply:
> So is the idea that formatter authors would provide a module with exactly this interface (specifically `use` and `register`) and not our `Hspec.JUnit.hs` module?

Basically yes, `use`, `register`, `formatter`, re-export of the used `Api` version and any additional primitives they choose to expose (here `formatterWith`).

One thing that's still up for discussion is weather

```hs
use :: SpecWith a
```

instead of

```hs
use :: SpecWith a -> SpecWith a
```

The earlier makes it clear that the formatter applies to the whole spec.

However, the later is more convenient for use in spec hooks. Specifically:

```hs
-- file: test/SpecHook.hs
module SpecHook where

import           Test.Hspec
import qualified MyFormatter

hook :: Spec -> Spec
hook spec = MyFormatter.use >> spec
```
vs
```hs
-- file: test/SpecHook.hs
module SpecHook where

import           Test.Hspec
import qualified MyFormatter

hook :: Spec -> Spec
hook = MyFormatter.use
```


> I ask because I think we can simplify `hspecJUnitWith` to something like,
> 
> 
> 
> ```hs
> 
> hspecWith base
> 
>   (if junitEnabled then JUnitFormatter.use else JUnitFormatter.register)
> 
>   spec
> 
> ```
👍

> And, at that point, the module's not particularly valuable.

The only reason to keep it would be for backwards compatibility.

Reply:
Here is the full example that will be included in the rendered docs btw https://github.com/hspec/hspec/tree/formatter-docs/doc/_includes/formatter/discover

Reply:
> ```haskell
> hspecWith base
>   (if junitEnabled then JUnitFormatter.use else JUnitFormatter.register)
>   spec
> ```

One more thing, instead of `JUNIT_ENABLED=1` you can use `HSPEC_FORMAT=junit` to select the formatter once it has been registered under that name.

Reply:
Oh wow, excellent.

Yeah, so with `HSPEC_FORMAT` available, we could reduce this package to only the `register` function. I know that's all that'd be necessary/useful to us (we'd set `HSPEC_FORMAT` on CI, so we don't even need `uses` ourselves). I'll have to learn a little more about other use cases to know if `uses` and `formatter`, etc would be worth us defining too.

I can't really speak to the difference in signature question for `uses`. I was surprised to see it as a `Spec -> Spec` and not a `Config -> Config`, which was more the angle we are using currently. But I guess that's because this pattern will extend to things that change more than just the `Config`?

I will mention that I don't personally see `m a` as indicating it applies to "the whole `m`" any more than `:: m a -> m a` does; it feels subject, but the latter almost indicates that more to me... it's a function that changes a [whole] `SpecWith a` to a new `SpecWith a`; I feel like `SpecWith a -> SpecWith a` is exactly the right signature. This may or may not be related, but I found `(modifyConfig (registerFormatter f) >>)` a bit dense. I'm curious what point in the design space made it come out like that, instead of simply `modifyConfig $ registerFormatter f`

I'm also not familiar with this first class "hooks" idea. We tend to manually write a `Main.hs` that imports a discovered `Spec.hs` and uses `hspecWith ... Spec.spec` for these cases that we want to modify things -- which is why this module has the shape it does today. Are hooks like that available today, or something you're introducing with this new idea?

So I guess the long-and-short of it is, this all looks great to me! My only wonder is how to move to it without dropping support for older hspec. Currently we have a perhaps-hacky-perhaps-clever compatibility story where the package functions against a range of hspec versions, but some of the nice-to-haves (like `--formatter=junit`) work or not based on the concrete version. I'll have to do some thinking about if/how to keep it like that with this new interface and the use of `HSPEC_FORMAT` -- or if we just go with a bucket of CPP or just drop support for older versions.

Reply:
> I'm also not familiar with this first class "hooks" idea. We tend to manually write a `Main.hs` that imports a discovered `Spec.hs` and uses `hspecWith ... Spec.spec` for these cases that we want to modify things -- which is why this module has the shape it does today. Are hooks like that available today, or something you're introducing with this new idea?

Spec hooks are available since `2.8.3`: https://hspec.github.io/hspec-discover.html#spec-hooks

The idea here is that you can apply `before` / `beforeAll` to a whole tests suite, or to some subtree of it.

As for `modifyConfig` (or `Formatter.register` for that matter), you could really do it anywhere. The top-level `SpecHook.hs` is just a convenient place to put it.

> I can't really speak to the difference in signature question for `uses`. I was surprised to see it as a `Spec -> Spec` and not a `Config -> Config`, which was more the angle we are using currently. But I guess that's because this pattern will extend to things that change more than just the `Config`?

The idea here is that ideally you never have to define you own `main` and just let `hspec-discover` do it's job. `test/SpecHook.hs` allows you to do any global setup / cleanup with `beforeAll` and friends and `modifyConfig` allows you to modify the `Config` in `SpecM` without having to resort to `hspecWith`.

> I will mention that I don't personally see `m a` as indicating it applies to "the whole `m`" any more than `:: m a -> m a` does; it feels subject, but the latter almost indicates that more to me... it's a function that changes a [whole] `SpecWith a` to a new `SpecWith a`; I feel like `SpecWith a -> SpecWith a` is exactly the right signature.

The thing is that you don't really apply it to a `spec`.  It does not matter where you use `modifyConfig`, anywhere will do.  Internally it's a `WriterT` over `Endo (Config -> Config)`.  That means that repeated calls to `modifyConfig` will be combined with `.`.

If you see something like

```haskell
spec = do
  modifyConfig (registerFormatter formatter)
  it "foo" $ do
    ...
  it "bar" $ do
    ...
```
then it's pretty clear that `modifyConfig` is only here for its side effect (here the side effect being that the config is modified).

With `register :: Spec -> Spec` on the other hand you can do something like this
```haskell
spec = do
  it "foo" $ do
    ...
  Formatter.register $ do
    it "bar" $ do
      ...
```
which could give the wrong impression that the formatter only applies to `it "bar"` somehow, and not `it "foo"`.

> This may or may not be related, but I found `(modifyConfig (registerFormatter f) >>)` a bit dense. I'm curious what point in the design space made it come out like that, instead of simply `modifyConfig $ registerFormatter f`

You mean why I defined `register` in terms of `(modifyConfig (registerFormatter f) >>)`?  The reason is that this form is more convenient for use in spec hooks, say 

```haskell
hook :: Spec -> Spec
hook = MyFormatter.register
```

Which can be easily combined with other hooks, e.g. 

```haskell
hook :: Spec -> Spec
hook = MyFormatter.register . aroundAll_ withEmptyEnv
```

Whereas if we define `register` as `modifyConfig $ registerFormatter f` (which then has type `Spec` instead of `Spec -> Spec`) a user would need to write

```haskell
hook :: Spec -> Spec
hook = (MyFormatter.register >>)
```
and
```haskell
hook :: Spec -> Spec
hook = (MyFormatter.register >>) . aroundAll_ withEmptyEnv
```
But maybe that's no so bad and would be less confusing?

> I'll have to learn a little more about other use cases to know if `uses` and `formatter`, etc would be worth us defining too.

Exposing the `formatter` allows for

1. a third party to extend the formatter
2. still using the `formatter` with `hspecWith` if for some reason somebody is inclined to do so

As for `use`, it's just an easy way to enable the formatter by default.  Arguably, it's probably not something that you would want for a `junit` formatter, but you probably want it for a formatter that is meant to be used during development.

> So I guess the long-and-short of it is, this all looks great to me! My only wonder is how to move to it without dropping support for older hspec. Currently we have a perhaps-hacky-perhaps-clever compatibility story where the package functions against a range of hspec versions, but some of the nice-to-haves (like `--formatter=junit`) work or not based on the concrete version. I'll have to do some thinking about if/how to keep it like that with this new interface and the use of `HSPEC_FORMAT` -- or if we just go with a bucket of CPP or just drop support for older versions.

Maybe just wait and do nothing for now and then at some point drop support for older `hspec` versions.  I would not use `CPP` for this.  Also regarding `HSPEC_FORMAT=junit`, it will fail if the formatter is not available (the same way as `--format=junit` would fail if there is no such formatter).

As a side note, there are not really any good reasons to use older versions of Hspec (new versions of Hspec still support GHC all the way back to `7.4.1` and the interface is reasonably stable, especially if you only use things exposed from the `hspec` package itself, say `Test.Hspec` etc.).